### ISSUE TITLE
Request: Expose comment nodes on AST from parseTemplate() to enable linter comments

### ISSUE THREAD

**Issue by JamesHenry** [angular/angular#41112]
# 🚀 feature request

### Relevant Package

 This feature request is for @angular/compiler

### Description

The template parser in https://github.com/angular-eslint/angular-eslint works by wrapping the @angular/compiler's `parseTemplate()` and doing some things with the output to teach ESLint what to do with it.

You can use comments in code to tell ESLint when to ignore a section of upcoming code, whether that's the next line, or whole sections delimited by Block comments.

We will be able to offer this same capability in Angular templates (both inline and external HTML files), if the `parseTemplate()` function starts returning an array of comment nodes on the root of the AST that it returns. The root comment nodes array can then be given to ESLint for attachment and everything should work as expected.

Here is example from code and the corresponding comment node structures we would like to emulate:

**source.ts:**
```ts
// a comment
import { foo } from 'bar'
/*
 multiline
*/
```

**Relative bit of AST:**
```jsonc
// ... root of AST
{
  //
  // ... more AST stuff here
  //
  "comments": [
    {
      "type": "Line",
      "value": " a comment",
      "start": 0,
      "end": 12,
      "range": [
        0,
        12
      ],
      "loc": {
        "start": {
          "line": 1,
          "column": 0
        },
        "end": {
          "line": 1,
          "column": 12
        }
      }
    },
    {
      "type": "Block",
      "value": "\n multiline\n",
      "start": 39,
      "end": 55,
      "range": [
        39,
        55
      ],
      "loc": {
        "start": {
          "line": 3,
          "column": 0
        },
        "end": {
          "line": 5,
          "column": 2
        }
      }
    }
  ]
}
```

### Describe the solution you'd like
 If you have a solution in mind, please describe it.

I would like my usage to be something like:

```diff
 const angularCompilerResult = parseTemplate(code, options.filePath, {
    preserveWhitespaces: true,
    preserveLineEndings: true,
+   comments: true
  });
```

And for it to return an array of comment Nodes at the root of the template AST, comparable to the code example above.

This will allow us to close: https://github.com/angular-eslint/angular-eslint/issues/97

### Describe alternatives you've considered
 Have you considered any alternative solutions or workarounds?

We could do an entirely separate parse of the template ourselves using separate tooling I guess, but obviously that would be very wasteful given the @angular/compiler is already doing that and we need the @angular/compiler's AST result anyway.

**Comment by petebacondarwin** [comment]
As far as I can tell the HTML parser already creates comment nodes in its output with equal importance to other HTML AST nodes. So this feature request would be to update this bit of code to attach these nodes to the "template AST":

https://github.com/angular/angular/blob/8d90c1ad97439ad657b14557c93e809c32ec31ac/packages/compiler/src/template_parser/template_parser.ts#L256-L258

**Comment by JamesHenry** [comment]
Thanks for pointing that out @petebacondarwin, in the meantime that has allowed me to hack at the compiler in my node_modules and write the actual conversion logic we will need once these nodes are available (converting `Comment` node to ESTree `Token`).

It's looking promising!


https://user-images.githubusercontent.com/900523/110327939-8ee3ea00-8012-11eb-82ed-2d42e5005eb5.mp4

**Comment by petebacondarwin** [comment]
Nice!

**Comment by JoostK** [comment]
This one was implemented in #41251 thanks to @JamesHenry himself, so closing this request.
